File: ts/project/.sdk/tm/rust/utility/result_basic.rs
use std::cell::RefCell;
use std::rc::Rc;

use crate::core::context::Context;
use crate::core::result::SdkResult;

pub fn result_basic_util(ctx: &Rc<Context>) -> Option<Rc<RefCell<SdkResult>>> {
    let response = ctx.response.borrow().clone();
    let result = ctx.result.borrow().clone();

    if let (Some(result), Some(response)) = (&result, &response) {
        let (status, status_text, resp_err) = {
            let r = response.borrow();
            (r.status, r.status_text.clone(), r.err.clone())
        };

        let mut res = result.borrow_mut();
        res.status = status;
        res.status_text = status_text.clone();

        if res.status >= 400 {
            let msg = format!("request: {}: {}", res.status, res.status_text);
            if let Some(prev) = &res.err {
                let prevmsg = prev.msg.clone();
                res.err = Some(ctx.make_error("request_status", &format!("{}: {}", prevmsg, msg)));
            } else {
                res.err = Some(ctx.make_error("request_status", &msg));
            }
        } else if let Some(rerr) = resp_err {
            res.err = Some(rerr);
        }
    }

    result
}

File: ts/project/.sdk/tm/c/utility/result_basic.c
// result_basic utility (mirrors utility/result_basic.rs).

#include "sdk.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static char* dup_str(const char* s) {
  if (!s) s = "";
  size_t n = strlen(s);
  char* d = (char*)malloc(n + 1);
  memcpy(d, s, n + 1);
  return d;
}

SdkResult* result_basic_util(Context* ctx) {
  Response* response = ctx->response;
  SdkResult* result = ctx->result;

  if (result && response) {
    result->status = response->status;
    free(result->status_text);
    result->status_text = dup_str(response->status_text);

    if (result->status >= 400) {
      char msg[512];
      snprintf(msg, sizeof(msg), "request: %lld: %s",
               (long long)result->status, result->status_text);
      if (result->err) {
        char buf[1024];
        snprintf(buf, sizeof(buf), "%s: %s", result->err->msg, msg);
        result->err = context_make_error(ctx, "request_status", buf);
      } else {
        result->err = context_make_error(ctx, "request_status", msg);
      }
    } else if (response->err) {
      result->err = response->err;
    }
  }

  return result;
}
